catalog/core/catalog-core-directorymonitor/src/main/java/org/codice/ddf/catalog/content/monitor/AsyncFileAlterationObserver.java

```diff
@@ -0,0 +1,311 @@
+
+      LOGGER.trace("Sending create Request for {}", entry.getName());
+
+      listener.onFileCreate(
+          entry.getFile(), new CompletionSynchronization(entry, this::commitCreate));
+    } else {
+      // Directories are always committed and added to the parent IF they
+      // don't already exist
+
+      File[] children = listFiles(entry.getFile());
+      for (File child : children) {
+        doCreate(new AsyncFileEntry(entry, child));
+      }
+
+      commitCreate(entry, true);
+    }
+  }
+
+  /**
+   * Callback to allow successful ContentFiles to commit themselves
+   *
+   * @param entry The AsyncFileEntry wrapping the file being listened to.
+   * @param success Boolean that shows if the task failed or completed successfully
+   */
+  private void commitCreate(AsyncFileEntry entry, boolean success) {
+
+    LOGGER.debug("commitCreate({},{}): Starting...", entry.getName(), success);
+    if (success) {
+      entry.commit();
+      entry.getParent().ifPresent(e -> e.addChild(entry));
+    }
+    removeFromProcessors(entry);
+  }
+
+  /**
+   * Fire directory/file change events to the registered listeners.
+   *
+   * @param entry The previous file system entry
+   */
+  private void doMatch(AsyncFileEntry entry) {
+
+    if (entry.hasChanged()) {
+      LOGGER.trace("{} has changed", entry.getName());
+      if (!entry.getFile().isDirectory()) {
+        if (!addToProcessors(entry)) {
+          return;
+        }
+        if (!entry.hasChanged()) {
+          //  Another thread just committed this. No work is needed.
+          removeFromProcessors(entry);
+          return;
+        }
+        LOGGER.trace("Sending Match Request for {}...", entry.getName());
+        listener.onFileChange(
+            entry.getFile(), new CompletionSynchronization(entry, this::commitMatch));
+      } else {
+        entry.commit();
+      }
+    }
+  }
+
+  /**
+   * Callback to allow successful {@link AsyncFileEntry}'s to commit their changes
+   *
+   * @param entry The AsyncFileEntry wrapping the file being listened to.
+   * @param success Boolean that shows if the task failed or completed successfully
+   */
+  private void commitMatch(AsyncFileEntry entry, boolean success) {
+    LOGGER.debug("commitMatch({},{}): Starting...", entry.getName(), success);
+    if (success) {
+      entry.commit();
+    }
+    removeFromProcessors(entry);
+  }
+
+  /**
+   * Fire directory/file delete events to the registered listeners.
+   *
+   * @param entry The file entry
+   */
+  private void doDelete(AsyncFileEntry entry) {
+    if (!addToProcessors(entry)) {
+      return;
+    }
+    if (entry.getFromParent() == null) {
+      //  If we don't exist in the parent another thread beat us here.
+      removeFromProcessors(entry);
+      return;
+    }
+
+    if (!entry.getFile().isDirectory() && !entry.isDirectory()) {
+      LOGGER.trace("Sending Delete Request for {}...", entry.getName());
+      listener.onFileDelete(
```

Comment:
✏️Could `listener` be `null` here?

Reply:
Potentially yes if the listener is removed after calling `checkAndNotify()` but before it finishes. Is this a case that we should account for?

Reply:
Fixed :+1: